File: ios/LessonsViewController.m
// Copyright 2018 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "LessonsViewController.h"

#import "Extensions/ProtobufExtensions.h"
#import "LessonsPageControl.h"
#import "ReviewItem.h"
#import "SubjectDetailsViewController.h"
#import "Tsurukame-Swift.h"
#import "UIView+SafeAreaInsets.h"

@interface LessonsViewController () <ReviewViewControllerDelegate>
@property(weak, nonatomic) IBOutlet LessonsPageControl *pageControl;
@property(weak, nonatomic) IBOutlet UIButton *backButton;
@property(nonatomic, readonly) NSArray<UIKeyCommand *> *keyCommands;
@end

@implementation LessonsViewController {
  TKMServices *_services;
  NSArray<ReviewItem *> *_items;

  UIPageViewController *_pageController;
  NSInteger _currentPageIndex;

  ReviewViewController *_reviewViewController;
}

- (void)setupWithServices:(id)services items:(NSArray<ReviewItem *> *)items {
  _services = services;
  _items = [items copy];
}

- (void)viewDidLoad {
  [super viewDidLoad];

  self.view.backgroundColor = TKMStyleColor.background;

  // Create the page controller.
  _pageController = [[UIPageViewController alloc]
      initWithTransitionStyle:UIPageViewControllerTransitionStyleScroll
        navigationOrientation:UIPageViewControllerNavigationOrientationHorizontal
                      options:nil];
  _pageController.dataSource = self;
  _pageController.delegate = self;

  // Set the subjects on the page control.
  NSMutableArray<TKMSubject *> *subjects = [NSMutableArray array];
  for (ReviewItem *item in _items) {
    TKMSubject *subject = [_services.dataLoader loadSubject:item.assignment.subjectId];
    if (!subject) {
      continue;
    }
    [subjects addObject:subject];
  }
  _pageControl.subjects = subjects;

  // Add it as a child view controller, below the back button.
  [self addChildViewController:_pageController];
  [self.view insertSubview:_pageController.view belowSubview:_backButton];
  [_pageController didMoveToParentViewController:self];

  // Hook up the page control.
  [_pageControl addTarget:self
                   action:@selector(pageChanged)
         forControlEvents:UIControlEventValueChanged];

  // Load the first page.
  [_pageController setViewControllers:@[ [self createViewControllerForIndex:0] ]
                            direction:UIPageViewControllerNavigationDirectionForward
                             animated:NO
                           completion:nil];
}

- (void)viewWillAppear:(BOOL)animated {
  [super viewWillAppear:animated];

  [SiriShortcutHelper.shared attachShortcutActivity:self
                                               type:SiriShortcutHelper.ShortcutTypeLessons];
}

- (void)viewDidLayoutSubviews {
  [super viewDidLayoutSubviews];

  CGRect safeArea = UIEdgeInsetsInsetRect(self.view.frame, self.view.tkm_safeAreaInsets);
  CGSize pageControlSize = [_pageControl sizeThatFits:CGSizeMake(self.view.frame.size.width, 0)];
  CGRect pageControlFrame = CGRectMake(CGRectGetMinX(safeArea),
                                       CGRectGetMaxY(safeArea) - pageControlSize.height,
                                       safeArea.size.width,
                                       pageControlSize.height);
  _pageControl.frame = pageControlFrame;
  [_pageControl setNeedsLayout];

  CGRect pageControllerFrame = self.view.frame;
  pageControllerFrame.size.height = pageControlFrame.origin.y;
  _pageController.view.frame = pageControllerFrame;
}

- (UIStatusBarStyle)preferredStatusBarStyle {
  return UIStatusBarStyleLightContent;
}

- (IBAction)didTapBackButton:(id)sender {
  [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - UIPageControl

- (void)pageChanged {
  NSInteger newPageIndex = _pageControl.currentPageIndex;
  if (newPageIndex == _currentPageIndex) {
    return;
  }

  UIViewController *vc = [self createViewControllerForIndex:newPageIndex];
  UIPageViewControllerNavigationDirection direction =
      (newPageIndex > _currentPageIndex) ? UIPageViewControllerNavigationDirectionForward
                                         : UIPageViewControllerNavigationDirectionReverse;
  [_pageController setViewControllers:@[ vc ] direction:direction animated:YES completion:nil];
  _currentPageIndex = newPageIndex;
}

#pragma mark - UIPageViewControllerDelegate

- (void)pageViewController:(UIPageViewController *)pageViewController
         didFinishAnimating:(BOOL)finished
    previousViewControllers:(NSArray<UIViewController *> *)previousViewControllers
        transitionCompleted:(BOOL)completed {
  NSInteger index = [self indexOfViewController:_pageController.viewControllers[0]];
  _pageControl.currentPageIndex = index;
  _currentPageIndex = index;
}

#pragma mark - UIPageViewControllerDataSource

- (UIViewController *)createViewControllerForIndex:(NSInteger)index {
  if (index == _items.count) {
    if (_reviewViewController == nil) {
      _reviewViewController =
          [self.storyboard instantiateViewControllerWithIdentifier:@"reviewViewController"];
      [_reviewViewController setupWithServices:_services
                                         items:_items
                                showMenuButton:NO
                            showSubjectHistory:NO
                                      delegate:self];
    }
    return _reviewViewController;
  } else if (index < 0 || index > _items.count) {
    return nil;
  }

  ReviewItem *item = _items[index];
  SubjectDetailsViewController *vc =
      [self.storyboard instantiateViewControllerWithIdentifier:@"subjectDetailsViewController"];
  [vc setupWithServices:_services
                subject:[_services.dataLoader loadSubject:item.assignment.subjectId]
              showHints:YES
         hideBackButton:YES
                  index:index];
  return vc;
}

- (NSInteger)indexOfViewController:(UIViewController *)viewController {
  if ([viewController.class isSubclassOfClass:[SubjectDetailsViewController class]]) {
    SubjectDetailsViewController *subjectDetailsViewController =
        (SubjectDetailsViewController *)viewController;
    return subjectDetailsViewController.index;
  } else if ([viewController.class isSubclassOfClass:[ReviewViewController class]]) {
    return _items.count;
  }
  return 0;
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController
       viewControllerAfterViewController:(UIViewController *)viewController {
  return [self createViewControllerForIndex:[self indexOfViewController:viewController] + 1];
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController
      viewControllerBeforeViewController:(UIViewController *)viewController {
  return [self createViewControllerForIndex:[self indexOfViewController:viewController] - 1];
}

#pragma mark - ReviewViewControllerDelegate

- (BOOL)reviewViewControllerAllowsCheatsForReviewItem:(ReviewItem *)reviewItem {
  return false;
}

- (void)reviewViewControllerFinishedAllReviewItems:(ReviewViewController *)reviewViewController {
  [reviewViewController.navigationController popToRootViewControllerAnimated:YES];
}

- (BOOL)reviewViewControllerAllowsCustomFonts {
  return false;
}

- (BOOL)reviewViewControllerShowsSuccessRate {
  return false;
}

#pragma mark - Keyboard navigation

- (BOOL)canBecomeFirstResponder {
  return true;
}

- (NSArray<UIKeyCommand *> *)keyCommands {
  // No keyboard nav on the quiz page, answer the quiz
  if (_pageControl.currentPageIndex == _items.count) {
    return @[];
  }

  return @[
    [UIKeyCommand keyCommandWithInput:UIKeyInputLeftArrow
                        modifierFlags:0
                               action:@selector(prevPage)
                 discoverabilityTitle:@"Previous"],
    [UIKeyCommand keyCommandWithInput:UIKeyInputRightArrow
                        modifierFlags:0
                               action:@selector(nextPage)
                 discoverabilityTitle:@"Next"],
    [UIKeyCommand keyCommandWithInput:@"\r" modifierFlags:0 action:@selector(nextPage)],
    [UIKeyCommand keyCommandWithInput:@" "
                        modifierFlags:0
                               action:@selector(playAudio)
                 discoverabilityTitle:@"Play reading"]

  ];
}

- (void)nextPage {
  if (_pageControl.currentPageIndex < [_items count]) {
    _pageControl.currentPageIndex += 1;
    [self pageChanged];
  }
}

- (void)prevPage {
  if (_pageControl.currentPageIndex > 0) {
    _pageControl.currentPageIndex -= 1;
    [self pageChanged];
  }
}

- (void)playAudio {
  UIViewController *vc = _pageController.viewControllers[0];
  if ([vc isKindOfClass:SubjectDetailsViewController.class]) {
    [(SubjectDetailsViewController *)vc playAudio];
  }
}

@end

File: ios/LessonsViewController.swift
// Copyright 2021 David Sansome
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

class LessonsViewController: UIViewController, UIPageViewControllerDataSource,
  UIPageViewControllerDelegate, ReviewViewControllerDelegate {
  private var services: TKMServices!
  private var items: [ReviewItem]!

  private var pageController: UIPageViewController!
  private var currentPageIndex = 0
  private var reviewViewController: ReviewViewController?

  @IBOutlet private var pageControl: LessonsPageControl!
  @IBOutlet private var backButton: UIButton!

  func setup(services: TKMServices, items: [ReviewItem]) {
    self.services = services
    self.items = items
  }

  override func viewDidLoad() {
    super.viewDidLoad()

    view.backgroundColor = TKMStyle.Color.background

    // Create the page controller.
    pageController = UIPageViewController(transitionStyle: .scroll,
                                          navigationOrientation: .horizontal, options: nil)
    pageController.delegate = self
    pageController.dataSource = self

    // Set the subjects on the page control.
    var subjects = [TKMSubject]()
    for item in items {
      if let subject = services.localCachingClient.getSubject(id: item.assignment.subjectID) {
        subjects.append(subject)
      }
    }
    pageControl.setSubjects(subjects)

    // Add it as a child view controller, below the back button.
    addChild(pageController)
    view.insertSubview(pageController.view, belowSubview: backButton)
    pageController.didMove(toParent: self)

    // Hook up the page control.
    pageControl.addTarget(self, action: #selector(pageChanged), for: .valueChanged)

    // Load the first page.
    if let vc = createViewController(index: 0) {
      pageController.setViewControllers([vc], direction: .forward,
                                        animated: false, completion: nil)
    }
  }

  override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    SiriShortcutHelper.shared.attachShortcutActivity(self, type: .lessons)
  }

  override func viewDidLayoutSubviews() {
    super.viewDidLayoutSubviews()

    let safeArea = view.frame.inset(by: view.tkm_safeAreaInsets)
    let pageControlSize = pageControl.sizeThatFits(CGSize(width: view.frame.size.width, height: 0))
    let pageControlFrame = CGRect(x: safeArea.minX, y: safeArea.maxY - pageControlSize.height,
                                  width: safeArea.size.width,
                                  height: pageControlSize.height)

    pageControl.frame = pageControlFrame
    pageControl.setNeedsLayout()

    var pageControllerFrame = view.frame
    pageControllerFrame.size.height = pageControlFrame.origin.y
    pageController.view.frame = pageControllerFrame
  }

  override var preferredStatusBarStyle: UIStatusBarStyle {
    .lightContent
  }

  @IBAction private func didTapBackButton(sender _: Any) {
    navigationController?.popViewController(animated: true)
  }

  // MARK: - UIPageControl

  @objc func pageChanged() {
    let newPageIndex = pageControl.currentPageIndex
    if newPageIndex == currentPageIndex {
      return
    }

    if let vc = createViewController(index: newPageIndex) {
      let direction: UIPageViewController
        .NavigationDirection = (newPageIndex > currentPageIndex) ? .forward : .reverse
      pageController.setViewControllers([vc], direction: direction, animated: true, completion: nil)
      currentPageIndex = newPageIndex
    }
  }

  // MARK: - UIPageViewControllerDelegate

  func pageViewController(_: UIPageViewController, didFinishAnimating _: Bool,
                          previousViewControllers _: [UIViewController],
                          transitionCompleted _: Bool) {
    let index = indexOf(viewController: pageController.viewControllers![0])
    pageControl.currentPageIndex = index
    currentPageIndex = index
  }

  // MARK: - UIPageViewControllerDataSource

  private func createViewController(index: Int) -> UIViewController? {
    if index == items.count {
      if reviewViewController == nil {
        reviewViewController = storyboard!
          .instantiateViewController(withIdentifier: "reviewViewController") as? ReviewViewController
        reviewViewController?.setup(services: services, items: items, showMenuButton: false,
                                    showSubjectHistory: false, delegate: self)
      }
      return reviewViewController
    } else if index < 0 || index > items.count {
      return nil
    }

    let item = items[index]
    if let subject = services.localCachingClient.getSubject(id: item.assignment.subjectID) {
      let vc = storyboard?
        .instantiateViewController(withIdentifier: "subjectDetailsViewController") as! SubjectDetailsViewController
      vc.setup(services: services, subject: subject, showHints: true, hideBackButton: true,
               index: index)
      return vc
    }
    return nil
  }

  private func indexOf(viewController: UIViewController) -> Int {
    if let viewController = viewController as? SubjectDetailsViewController {
      return viewController.index
    } else if viewController is ReviewViewController {
      return items.count
    }
    return 0
  }

  func pageViewController(_: UIPageViewController,
                          viewControllerAfter viewController: UIViewController)
    -> UIViewController? {
    createViewController(index: indexOf(viewController: viewController) + 1)
  }

  func pageViewController(_: UIPageViewController,
                          viewControllerBefore viewController: UIViewController)
    -> UIViewController? {
    createViewController(index: indexOf(viewController: viewController) - 1)
  }

  // MARK: - ReviewViewControllerDelegate

  func reviewViewControllerAllowsCheats(forReviewItem _: ReviewItem) -> Bool {
    false
  }

  func reviewViewControllerFinishedAllReviewItems(_ reviewViewController: ReviewViewController) {
    reviewViewController.navigationController?.popToRootViewController(animated: true)
  }

  func reviewViewControllerAllowsCustomFonts() -> Bool {
    false
  }

  func reviewViewControllerShowsSuccessRate() -> Bool {
    false
  }

  // MARK: - Keyboard navigation

  override var canBecomeFirstResponder: Bool {
    true
  }

  override var keyCommands: [UIKeyCommand]? {
    // No keyboard nav on the quiz page, answer the quiz.
    if pageControl.currentPageIndex == items.count {
      return []
    }

    return [
      UIKeyCommand(input: UIKeyCommand.inputLeftArrow, modifierFlags: [],
                   action: #selector(prevPage),
                   discoverabilityTitle: "Previous"),
      UIKeyCommand(input: "a", modifierFlags: [], action: #selector(prevPage)),
      UIKeyCommand(input: UIKeyCommand.inputRightArrow, modifierFlags: [],
                   action: #selector(nextPage),
                   discoverabilityTitle: "Next"),
      UIKeyCommand(input: "d", modifierFlags: [], action: #selector(nextPage)),
      UIKeyCommand(input: "\r", modifierFlags: [], action: #selector(nextPage)),
      UIKeyCommand(input: " ", modifierFlags: [], action: #selector(playAudio),
                   discoverabilityTitle: "Play reading"),
      UIKeyCommand(input: "j", modifierFlags: [], action: #selector(playAudio)),
      UIKeyCommand(input: "q", modifierFlags: [], action: #selector(jumpToQuiz),
                   discoverabilityTitle: "Jump to quiz"),
    ]
  }

  @objc func prevPage() {
    if pageControl.currentPageIndex > 0 {
      pageControl.currentPageIndex -= 1
      pageChanged()
    }
  }

  @objc func nextPage() {
    if pageControl.currentPageIndex < items.count {
      pageControl.currentPageIndex += 1
      pageChanged()
    }
  }

  @objc func jumpToQuiz() {
    pageControl.currentPageIndex = items.count
    pageChanged()
  }

  @objc func playAudio() {
    if let vc = pageController.viewControllers?.first as? SubjectDetailsViewController {
      vc.playAudio()
    }
  }
}
